src/os/bluestore/Btree2Allocator.h

```diff
@@ -0,0 +1,425 @@
+    }
+    ceph_assert(idx < num_size_buckets);
+    return idx;
+  }
+};
+
+class ChunkCache {
+public:
+  class LocklessBuf {
+    std::atomic<size_t> ref = 0;
+    size_t count = 0;
+    std::vector<uint32_t> data;
+  public:
+    void init(size_t size) {
+      ceph_assert(data.size() == 0);
+      data.resize(size);
+    }
+    bool maybe_put(uint32_t& v) {
+      bool done = ++ref == 1 && count < data.size();
+      if (done) {
+        data[count++] = v;
+      }
+      --ref;
+      return done;
+    }
+    bool maybe_get(uint32_t& v) {
+      bool done = ++ref == 1 && count > 0;
+      if (done) {
+        v = data[--count];
+      }
+      --ref;
+      return done;
+    }
+    void foreach(std::function<void(uint32_t)> notify) {
+      for (size_t i = 0; i < count; i++) {
+        notify(data[i]);
+      }
+    }
+  };
+
+  static const size_t base_bits = 12; // 4K
+  static const uint64_t base_mask = (1ull << base_bits) - 1;
+  static const size_t num_buckets = 16; // [4K, 8K, 12K, ... 64K] - 16 buckets total
+  static const size_t num_chunks = 16;
+  ChunkCache() {
+    for (size_t i = 0; i < buckets.size(); i++) {
+      buckets[i].init(num_chunks);
+    }
+  }
+  bool maybe_put(uint64_t offset, uint64_t len) {
+    if (!lock.try_lock_shared()) {
+      return false;
+    }
+    bool ret = false;
+    ceph_assert(!(offset & base_mask));
+    ceph_assert(!(len & base_mask));
+    // permit caching chunks with offsets fitting (without LSBs)
+    // into 32 - bit value
+    offset = offset >> base_bits;
+    size_t idx = len >> base_bits;
+    if (offset <= std::numeric_limits<uint32_t>::max() &&
+      idx < buckets.size()) {
+      uint32_t o = offset;
+      ret = buckets[idx].maybe_put(o);
+    }
+    lock.unlock_shared();
+    return ret;
+  }
+  bool maybe_get(uint64_t* offset, uint64_t len) {
+    if (!lock.try_lock_shared()) {
+      return false;
+    }
+    bool ret = false;
+    ceph_assert(offset);
+    ceph_assert(!(len & base_mask));
+    size_t idx = len >> base_bits;
+    if (idx < buckets.size()) {
+      uint32_t o = 0;
+      ret = buckets[idx].maybe_get(o);
+      if (ret) {
+        *offset = uint64_t(o) << base_bits;
+        ++hits;
+      }
+    }
+    lock.unlock_shared();
+    return ret;
+  }
+  size_t get_hit_count() const {
+    return hits.load();
+  }
+  void foreach(std::function<void(uint64_t offset, uint64_t length)> notify) {
+    std::unique_lock _lock(lock);
+    for (uint64_t i = 0; i < buckets.size(); i++) {
+      auto cb = [&](uint32_t o) {
+        notify(uint64_t(o) << base_bits, i << base_bits);
+      };
+      buckets[i].foreach(cb);
+    }
+  }
+private:
+  std::array<LocklessBuf, num_buckets> buckets;
+  std::atomic<size_t> hits = 0;
+  ceph::shared_mutex lock{
+    ceph::make_shared_mutex(std::string(), false, false, false)
+  };
+};
+
+/*
+ * class Btree2Allocator
+ *
+ *
+ */
+class Btree2Allocator : public Allocator,
+                        public SizeBucketCollectionBase<12, 14> {
```

Comment:
Why we inherit from SizeBucketCollectionBase, not encapsulate it? <12, 14> does not depend on anything.